src/webgpu/util/constants.ts

```diff
@@ -24,6 +24,67 @@ export const kBit = {
     max: 0xffff_ffff,
   },
 
+  // Limits of f64
+  // Have to be stored as a BigInt hex value, since number is a f64 internally,
+  // so 64-bit hex values are not guaranteed to be precisely representable.
+  f64: {
+    positive: {
+      min: BigInt(0x0010_0000_0000_0000n),
+      max: BigInt(0x7fef_ffff_ffff_ffffn),
+      zero: BigInt(0x0000_0000_0000_0000n),
+      nearest_max: BigInt(0x7fef_ffff_ffff_fffen),
+      less_than_one: BigInt(0x3fef_ffff_ffff_ffffn),
+      pi: {
+        whole: BigInt(0x4009_21fb_5444_2d18n),
+        three_quarters: BigInt(0x4002_d97c_7f33_21d2n),
+        half: BigInt(0x3ff9_21fb_5444_2d18n),
+        third: BigInt(0x3ff0_c152_382d_7366n),
+        quarter: BigInt(0x3fe9_21fb_5444_2d18n),
+        sixth: BigInt(0x3fe0_c152_382d_7366n),
+      },
+      e: BigInt(0x4005_bf0a_8b14_5769n),
+    },
+    negative: {
+      max: BigInt(0x8010_0000_0000_0000n),
+      min: BigInt(0xffef_ffff_ffff_ffffn),
+      zero: BigInt(0x8000_0000_0000_0000n),
+      nearest_min: BigInt(0xffef_ffff_ffff_fffen),
+      less_than_one: BigInt(0xbfef_ffff_ffff_ffffn),
+      pi: {
+        whole: BigInt(0xc009_21fb_5444_2d18n),
+        three_quarters: BigInt(0xc002_d97c_7f33_21d2n),
+        half: BigInt(0xbff9_21fb_5444_2d18n),
+        third: BigInt(0xbff0_c152_382d_7366n),
+        quarter: BigInt(0xbfe9_21fb_5444_2d18n),
+        sixth: BigInt(0xbfe0_c152_382d_7366n),
+      },
+    },
+    subnormal: {
+      positive: {
+        min: BigInt(0x0000_0000_0000_0001n),
+        max: BigInt(0x000f_ffff_ffff_ffffn),
+      },
+      negative: {
+        max: BigInt(0x8000_0000_0000_0001n),
+        min: BigInt(0x800f_ffff_ffff_ffffn),
+      },
+    },
+    nan: {
+      negative: {
+        s: BigInt(0xfff0_0000_0000_0001n),
+        q: BigInt(0xfff8_0000_0000_0001n),
```

Comment:
> Quiet and signalling nan differ in the top bit of the significand.

That's an implementation detail. This is not specified by any specification that I'm aware of. We shouldn't be relying on it.

Reply:
It's something I remember "forever", so I checked IEEE 754 2008.  Definitely something I didn't read from that spec before, but it says:

> 6.2.1 NaN encodings in binary formats

> ....  A quiet NaN bit
string should be encoded with the first bit (d1) of the trailing significand field T being 1. A signaling NaN
bit string should be encoded with the first bit of the trailing significand field being 0. If the first bit of the
trailing significand field is 0, some other bit of the trailing significand field must be non-zero to distinguish
the NaN from infinity. 

In an earlier diagram d1 is shown as the MSB of the significand.

It also generally says that the remaining bits of the significand should encode diagnostic information, and that NaNs should be passed on through operations, to propagate that diagnostic info.


https://en.wikipedia.org/wiki/NaN#Encoding says that many processors follow this convention.
Amusingly, PA-RISC and old MIPS CPUs flip the sense of that top bit.  It also gives a rationale for following IEEE (you can convert signalling to quiet by blindly setting the top bit to 1.)

All that said, WGSL does not specify this detail.  So we do have to be careful about overconstraining the tests.

What this *is* good for is to introduce diversity in the test inputs, because it's a likely thing to trigger differing behaviour, in the CPU side, and in GPUs that try harder to do things the IEEE way.

Reply:
Humm, I take it back. I had read previously that this signalling bit was undefined. Clearly that was wrong. My apologies.